src/Time/Actions/LimitTimeStepToExpirationTimes.hpp

```diff
@@ -0,0 +1,175 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines action AdvanceTime
+
+#pragma once
+
+#include <cmath>
+#include <iomanip>
+#include <limits>
+#include <optional>
+#include <tuple>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "Domain/Structure/ElementId.hpp"
+#include "Parallel/AlgorithmExecution.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "Time/Tags.hpp"
+#include "Time/Time.hpp"
+#include "Time/TimeStepId.hpp"
+#include "Time/TimeSteppers/TimeStepper.hpp"
+#include "Time/Utilities.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/FractionUtilities.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+#include "Evolution/EventsAndDenseTriggers/Tags.hpp"
+#include "Parallel/Printf.hpp"
+#include "Time/Tags.hpp"
+
+/// \cond
+namespace Parallel {
+template <typename Metavariables>
+class GlobalCache;
+}  // namespace Parallel
+namespace Tags {
+template <typename Tag>
+struct Next;
+}  // namespace Tags
+// IWYU pragma: no_forward_declare db::DataBox
+/// \endcond
+
+namespace Actions {
+/// \ingroup ActionsGroup
```

Comment:
Can this become a `SlabChooser`? Given the requirement that one uses `Slabchoosers` anyway, and that there is no check enforcing that (as far as I can tell), it seems like the safest thing to do is to have this be done as a `SlabChooser`. @wthrowe any thoughts on this?

Reply:
I think it can be done at the cost of complicating the action loop some more.  The annoying part here is that Adams-Bashforth and Runge-Kutta time steppers do dense output differently.  AB requires the step size to change before dense output, while RK doesn't care.  (The AB behavior might change to match RK at some point, but not in the near future.)  This computation has to happen after dense output, but isn't used for AB.  So I think it would work if we had two versions of the ChangeSlabSize action, one that only ran for AB and one that only ran for RK, and put them on either side of dense output.

ChangeSlabSize is currently done well before dense output, but I think it commutes with all the actions in between in GTS mode.  (Not true in LTS mode.)

Reply:
Okay, thank you! I guess for me it's not clear if having two `ChangeSlabsize` actions is more complicated than having to add this action anyway. I'm fine either way, with a slight preference to having it be a `SlabChooser`, but I think you should make the decision :)